validate_output.java

```diff
@@ -0,0 +1,124 @@
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+import java.io.BufferedReader;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.regex.Pattern;
+import java.util.stream.IntStream;
+
+public class validate_output {
+    private static final Pattern numRegex = Pattern.compile("-?\\d\\d?\\.\\d");
+
+    record NameAndStats(String name, int[] stats) {
+    }
+
+    public static void main(String[] args) throws IOException {
+        boolean lenient;
+        int argIndex;
+        if (args[0].equals("--lenient")) {
+            lenient = true;
+            argIndex = 1;
+        }
+        else {
+            lenient = false;
+            argIndex = 0;
+        }
+        var fnameExpected = args[argIndex];
+        var fnameActual = args[argIndex + 1];
+        var entriesExpected = entries(fnameExpected);
+        var entriesActual = entries(fnameActual);
+        if (entriesExpected.length != entriesActual.length) {
+            System.out.printf("Files don't match in the number of entries: expected %,d, actual %,d%n", entriesExpected.length, entriesActual.length);
+            System.exit(1);
+        }
+        for (int i = 0; i < entriesExpected.length; i++) {
+            var nameAndStatsExpected = parseEntry(entriesExpected[i], i, fnameExpected);
+            var nameAndStatsActual = parseEntry(entriesActual[i], i, fnameActual);
+            int[] statsExpected = nameAndStatsExpected.stats;
+            int[] statsActual = nameAndStatsActual.stats;
+            var mismatch = statsExpected[0] != statsActual[0] || statsExpected[2] != statsActual[2];
+            if (!mismatch) {
+                var avgExpected = statsExpected[1];
+                var avgActual = statsActual[1];
+                mismatch = avgExpected != avgActual;
+                if (lenient) {
+                    mismatch &= avgExpected - avgActual != 1;
+                }
+            }
+            if (mismatch) {
+                System.out.printf("Values don't match in entry #%,d: expected \"%s\", actual \"%s\"%n", i, entriesExpected[i], entriesActual[i]);
+                System.exit(1);
+            }
+        }
+    }
+
+    private static String[] entries(String fname) throws IOException {
+        var content = new BufferedReader(new FileReader(fname)).readLine();
+        if (!content.startsWith("{") || !content.endsWith("}")) {
+            System.out.printf("%s isn't in the format \"{entries}\"%n", fname);
+            System.exit(1);
+        }
+        content = content.substring(1, content.length() - 1);
+        return content.split(",");
```

Comment:
The whole thing kind of shows the importance of using a proper format https://github.com/gunnarmorling/1brc/issues/14

See also https://github.com/gunnarmorling/1brc/blob/a0f826c2e4441c93b4c6cf54df77793c03f2cc67/tocsv.sh#L29-L31